core/chains/evm/gas/rollups/zkSync_l1_oracle.go

```diff
@@ -0,0 +1,256 @@
+package rollups
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils"
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/config/chaintype"
+)
+
+// Reads L2-specific precompiles and caches the l1GasPrice set by the L2.
+type zkSyncL1Oracle struct {
+	services.StateMachine
+	client     l1OracleClient
+	pollPeriod time.Duration
+	logger     logger.SugaredLogger
+	chainType  chaintype.ChainType
+
+	systemContextAddress  string
+	gasPerPubdataMethod   string
+	gasPerPubdataSelector string
+	l2GasPriceMethod      string
+	l2GasPriceSelector    string
+
+	l1GasPriceMu sync.RWMutex
+	l1GasPrice   priceEntry
+
+	chInitialised chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+}
+
+const (
+	// SystemContextAddress is the address of the "Precompiled contract that calls that holds the current gas per pubdata byte"
+	// https://sepolia.explorer.zksync.io/address/0x000000000000000000000000000000000000800b#contract
+	SystemContextAddress = "0x000000000000000000000000000000000000800B"
+
+	// ZksyncGasInfo_GetL2GasPerPubDataBytes is the a hex encoded call to:
+	// function gasPerPubdataByte() external view returns (uint256 gasPerPubdataByte);
+	SystemContext_gasPerPubdataByteMethod = "gasPerPubdataByte"
+	ZksyncGasInfo_getGasPerPubdataByteL2  = "0x7cb9357e"
+
+	// ZksyncGasInfo_GetL2GasPrice is the a hex encoded call to:
+	// `function gasPrice() external view returns (uint256);`
+	SystemContext_gasPriceMethod = "gasPrice"
+	ZksyncGasInfo_getGasPriceL2  = "0xfe173b97"
+)
+
+func NewZkSyncL1GasOracle(lggr logger.Logger, ethClient l1OracleClient) *zkSyncL1Oracle {
+	var systemContextAddress, gasPerPubdataMethod, gasPerPubdataSelector, l2GasPriceMethod, l2GasPriceSelector string
+
+	systemContextAddress = SystemContextAddress
+	gasPerPubdataMethod = SystemContext_gasPerPubdataByteMethod
+	gasPerPubdataSelector = ZksyncGasInfo_getGasPerPubdataByteL2
+
+	l2GasPriceMethod = SystemContext_gasPriceMethod
+	l2GasPriceSelector = ZksyncGasInfo_getGasPriceL2
```

Comment:
Couldn't we just set the fields in the struct directly?

Reply:
yea good point, just followed the other oracles implementation, and just realised its unnecessary

Reply:
Noticed that in the others too. Not sure why they were written like that tbh. Likely in need of a clean up.

Reply:
yeah, I cleaned up zkSync anyway, good catch